client/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -0,0 +1,476 @@
+ * failed (the more failures, the later they will be retried). In case of failures all of the alive nodes (or dead nodes that
+ * deserve a retry) are retried till one responds or none of them does, in which case an {@link IOException} will be thrown.
+ *
+ * Requests can be traced by enabling trace logging for "tracer". The trace logger outputs requests and responses in curl format.
+ */
+public final class RestClient implements Closeable {
+
+    private static final Log logger = LogFactory.getLog(RestClient.class);
+    public static ContentType JSON_CONTENT_TYPE = ContentType.create("application/json", Consts.UTF_8);
+
+    private final CloseableHttpClient client;
+    //we don't rely on default headers supported by HttpClient as those cannot be replaced, plus it would get hairy
+    //when we create the HttpClient instance on our own as there would be two different ways to set the default headers.
+    private final Header[] defaultHeaders;
+    private final long maxRetryTimeout;
+    private final AtomicInteger lastHostIndex = new AtomicInteger(0);
+    private volatile Set<HttpHost> hosts;
+    private final ConcurrentMap<HttpHost, DeadHostState> blacklist = new ConcurrentHashMap<>();
+    private volatile FailureListener failureListener = new FailureListener();
+
+    private RestClient(CloseableHttpClient client, long maxRetryTimeout, Header[] defaultHeaders, HttpHost[] hosts) {
+        this.client = client;
+        this.maxRetryTimeout = maxRetryTimeout;
+        this.defaultHeaders = defaultHeaders;
+        setHosts(hosts);
+    }
+
+    /**
+     * Replaces the hosts that the client communicates with.
+     * @see HttpHost
+     */
+    public synchronized void setHosts(HttpHost... hosts) {
+        if (hosts == null || hosts.length == 0) {
+            throw new IllegalArgumentException("hosts must not be null nor empty");
+        }
+        Set<HttpHost> httpHosts = new HashSet<>();
+        for (HttpHost host : hosts) {
+            Objects.requireNonNull(host, "host cannot be null");
+            httpHosts.add(host);
+        }
+        this.hosts = Collections.unmodifiableSet(httpHosts);
+        this.blacklist.clear();
+    }
+
+    /**
+     * Sends a request to the elasticsearch cluster that the current client points to.
+     * Selects a host out of the provided ones in a round-robin fashion. Failing hosts are marked dead and retried after a certain
+     * amount of time (minimum 1 minute, maximum 30 minutes), depending on how many times they previously failed (the more failures,
+     * the later they will be retried). In case of failures all of the alive nodes (or dead nodes that deserve a retry) are retried
+     * till one responds or none of them does, in which case an {@link IOException} will be thrown.
+     *
+     * @param method the http method
+     * @param endpoint the path of the request (without host and port)
```

Comment:
this path doesn't contain query_string params though, that is why I called it endpoint. params come in as a map. What did you have in mind?